src/macaron/repo_finder/repo_verifier.py

```diff
@@ -0,0 +1,300 @@
+            ):
+                return RepositoryVerificationResult(
+                    status=RepositoryVerificationStatus.PASSED, reason="git_ns", build_tool=self.build_tool
+                )
+
+        return RepositoryVerificationResult(
+            # not necessarily a fail, because many projects use maven group ids other than their repo domain.
+            status=RepositoryVerificationStatus.UNKNOWN,
+            reason="git_ns_mismatch",
+            build_tool=self.build_tool,
+        )
+
+    def _verify_maven(self) -> RepositoryVerificationResult:
+        # TODO: check other pom files. think about how to decide in case of contradicting evidence
+        # check if repo contains pom.xml
+        pom_file = self._bfs_walk(Path(self.claimed_repo_fs), "pom.xml")
+        if pom_file is None:
+            logger.debug("Could not find any pom.xml in the repository: %s", self.claimed_repo_url)
+            return RepositoryVerificationResult(
+                status=RepositoryVerificationStatus.UNKNOWN, reason="no_pom", build_tool=self.build_tool
+            )
+
+        pom_content = pom_file.read_text()
+        pom_root = parse_pom_string(pom_content)
+
+        if not pom_root:
+            logger.debug("Could not parse pom.xml: %s", pom_file.as_posix())
+            return RepositoryVerificationResult(
+                status=RepositoryVerificationStatus.UNKNOWN, reason="not_parsed_pom", build_tool=self.build_tool
+            )
+
+        # find the group id in the pom (project/groupId)
+        pom_group_id_elem = next((ch for ch in pom_root if ch.tag.endswith("}groupId")), None)
+        if pom_group_id_elem is None or pom_group_id_elem.text is None:
+            logger.debug("Could not find groupId in pom.xml: %s", pom_file)
+            return RepositoryVerificationResult(
+                status=RepositoryVerificationStatus.UNKNOWN, reason="no_group_id_in_pom", build_tool=self.build_tool
+            )
+
+        pom_group_id = pom_group_id_elem.text.strip()
+        if not self._same_group(pom_group_id, self.namespace):
+            logger.debug("Group id in pom.xml does not match the provided group id: %s", pom_file)
+            return RepositoryVerificationResult(
+                status=RepositoryVerificationStatus.FAILED, reason="group_id_mismatch", build_tool=self.build_tool
+            )
+
+        return RepositoryVerificationResult(
+            status=RepositoryVerificationStatus.PASSED, reason="group_id_match", build_tool=self.build_tool
+        )
+
+    @staticmethod
+    def _is_valid_maven_group_id(group_id: str) -> bool:
+        pattern = r"^[a-zA-Z][a-zA-Z0-9-]*\.([a-zA-Z][a-zA-Z0-9-]*\.)*[a-zA-Z][a-zA-Z0-9-]*[a-zA-Z0-9]$"
+        return re.match(pattern, group_id) is not None
+
+    def _verify_gradle(self) -> RepositoryVerificationResult:
+        # check if repo contains gradle.properties
+        def _extract_group_id_from_properties() -> str | None:
```

Comment:
This and the following two functions are very similar. I would suggest separating out the shared the code and reusing it.